src/solver/td3.ml

```diff
@@ -346,16 +387,16 @@ module Base =
               if tracing then trace "sol2" "solve still unstable %a" S.Var.pretty_trace x;
               (solve[@tailcall]) x Widen
             ) else (
-              if term && phase = Widen && HM.mem wpoint x then ( (* TODO: or use wp? *)
+              if term && phase = Widen && should_widen x then ( (* TODO: or use wp? *)
                 if tracing then trace "sol2" "solve switching to narrow %a" S.Var.pretty_trace x;
                 if tracing then trace "sol2" "stable remove %a" S.Var.pretty_trace x;
                 HM.remove stable x;
                 HM.remove superstable x;
                 Hooks.stable_remove x;
                 (solve[@tailcall]) ~reuse_eq:eqd x Narrow
```

Comment:
Is it right to swap `HM.mem wpoint` for `should_widen` here?
This case is taken when in the widening phase the value has stabilized and thus should be switched to narrowing.
If I understand correctly, this now means that if all the widening has isn't used up at `x` (because `should_widen` requires it to be 0), then the solver doesn't switch to narrowing at all.

Previously the notions of "widening point" and "should apply widening now" more-or-less coincided, but with gas they no longer do. I think places like this that check `HM.mem wpoint` but don't actually call `widen` based on that should still check for "widening point", not "should apply widening now".

Reply:
I think it should be appropriate to use `should_widen` here. If x is a widening point with gas, that means it was never widened. If it was not widened, then narrowing would be pointless.

On the other hand, using `HM.mem wpoint` would also not be harmful either, nor would narrowing cause a large overhead if x does not change. I can change it to a membership check, if you prefer

Reply:
Ah, I guess you are right, narrowing could still have an effect due to some non-monotonicity.